# Traffic Sign Classification with Keras

Keras exists to make coding deep neural networks simpler. To demonstrate just how easy it is, you’re going to use Keras to build a convolutional neural network in a few dozen lines of code.

You’ll be connecting the concepts from the previous lessons to the methods that Keras provides.

## Dataset

The network you'll build with Keras is similar to the example that you can find in Keras’s GitHub repository that builds out a [convolutional neural network for MNIST](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py). 

However, instead of using the [MNIST](http://yann.lecun.com/exdb/mnist/) dataset, you're going to use the [German Traffic Sign Recognition Benchmark](http://benchmark.ini.rub.de/?section=gtsrb&subsection=news) dataset that you've used previously.

You can download pickle files with sanitized traffic sign data here.

## Overview

Here are the steps you'll take to build the network:

1. First load the training data and do a train/validation split.
2. Preprocess data.
3. Build a feedforward neural network to classify traffic signs.
4. Build a convolutional neural network to classify traffic signs.
5. Evaluate performance of final neural network on testing data.

Keep an eye on the network’s accuracy over time. Once the accuracy reaches the 98% range, you can be confident that you’ve built and trained an effective model.

In [1]:
import pickle
import numpy as np
from sklearn.model_selection import train_test_split
import math

## Load the Data

Start by importing the data from the pickle file.

In [2]:
# TODO: Implement load the data here.
training_file = 'train.p'
testing_file = 'test.p'

with open(training_file, mode='rb') as f_train:
    train = pickle.load(f_train)
with open(testing_file, mode='rb') as f_test:
    test = pickle.load(f_test)

# Split validation test set from train test set
X_train, X_val, y_train, y_val = train_test_split(train["features"], train["labels"], random_state=0, test_size=0.33)

# Test data set
X_test, y_test = test["features"], test["labels"]

## Validate the Network
Split the training data into a training and validation set.

Measure the [validation accuracy](https://keras.io/models/sequential/) of the network after two training epochs.

Hint: [Use the `train_test_split()` method](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) from scikit-learn.

In [4]:
# TODO: Use `train_test_split` here.

X_train, X_val, y_train, y_val = train_test_split(train['features'], train['labels'], random_state=0, test_size=0.33)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)

(26270, 32, 32, 3) (26270,)
(12939, 32, 32, 3) (12939,)


In [5]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert(X_train.shape[0] == y_train.shape[0]), "The number of images is not equal to the number of labels."
assert(X_train.shape[1:] == (32,32,3)), "The dimensions of the images are not 32 x 32 x 3."
assert(X_val.shape[0] == y_val.shape[0]), "The number of images is not equal to the number of labels."
assert(X_val.shape[1:] == (32,32,3)), "The dimensions of the images are not 32 x 32 x 3."

## Preprocess the Data

Now that you've loaded the training data, preprocess the data such that it's in the range between -0.5 and 0.5.

In [6]:
# TODO: Implement data normalization here.
X_train = X_train.astype('float32')
X_val = X_val.astype('float32')
X_test = X_train.astype('float32')

X_train = X_train/255. - .5
X_val = X_val/255. - .5
X_test = X_test/255 - .5


In [7]:
# STOP: Do not change the tests below. Your implementation should pass these tests. 
assert(math.isclose(np.min(X_train), -0.5, abs_tol=1e-5) and math.isclose(np.max(X_train), 0.5, abs_tol=1e-5)), "The range of the training data is: %.1f to %.1f" % (np.min(X_train), np.max(X_train))
assert(math.isclose(np.min(X_val), -0.5, abs_tol=1e-5) and math.isclose(np.max(X_val), 0.5, abs_tol=1e-5)), "The range of the validation data is: %.1f to %.1f" % (np.min(X_val), np.max(X_val))

## Build a Two-Layer Feedfoward Network

The code you've written so far is for data processing, not specific to Keras. Here you're going to build Keras-specific code.

Build a two-layer feedforward neural network, with 128 neurons in the fully-connected hidden layer. 

To get started, review the Keras documentation about [models](https://keras.io/models/sequential/) and [layers](https://keras.io/layers/core/).

The Keras example of a [Multi-Layer Perceptron](https://github.com/fchollet/keras/blob/master/examples/mnist_mlp.py) network is similar to what you need to do here. Use that as a guide, but keep in mind that there are a number of differences.

In [8]:
# TODO: Build a two-layer feedforward neural network with Keras here.

from keras.models import Sequential
from keras.layers import Dense, Input, Activation

model = Sequential()
hidden1 = model.add(Dense(128, activation='relu', input_dim=3072))
output = model.add(Dense(43, activation='softmax'))

Using TensorFlow backend.


In [9]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
dense_layers = []
for l in model.layers:
    if type(l) == Dense:
        dense_layers.append(l)
assert(len(dense_layers) == 2), "There should be 2 Dense layers."
d1 = dense_layers[0]
d2 = dense_layers[1]
assert(d1.input_shape == (None, 3072))
assert(d1.output_shape == (None, 128))
assert(d2.input_shape == (None, 128))
assert(d2.output_shape == (None, 43))

last_layer = model.layers[-1]
assert(last_layer.activation.__name__ == 'softmax'), "Last layer should be softmax activation, is {}.".format(last_layer.activation.__name__)

In [10]:
# Debugging
for l in model.layers:
    print(l.name, l.input_shape, l.output_shape, l.activation)

dense_1 (None, 3072) (None, 128) <function relu at 0x12ee43488>
dense_2 (None, 128) (None, 43) <function softmax at 0x12ee43268>


## Train the Network
Compile and train the network for 2 epochs. [Use the `adam` optimizer, with `categorical_crossentropy` loss.](https://keras.io/models/sequential/)

Hint 1: In order to use categorical cross entropy, you will need to [one-hot encode the labels](https://github.com/fchollet/keras/blob/master/keras/utils/np_utils.py).

Hint 2: In order to pass the input images to the fully-connected hidden layer, you will need to [reshape the input](https://github.com/fchollet/keras/blob/master/examples/mnist_mlp.py).

Hint 3: Keras's `.fit()` method returns a `History.history` object, which the tests below use. Save that to a variable named `history`.

In [15]:
# TODO: Compile and train the model here.

# model.summary()

from keras.utils import np_utils

# one-hot encode the labels

Y_train = np_utils.to_categorical(y_train, 43)
Y_val = np_utils.to_categorical(y_val, 43)

# reshape the input
X_train_flat = X_train.reshape(-1, 32*32*3)
X_val_flat = X_val.reshape(-1, 32*32*3)

print(X_train_flat.shape, Y_train.shape)
print(X_val_flat.shape, Y_val.shape)

model.compile(optimizer="adam", loss="categorical_crossentropy", metrics=['accuracy'])
history = model.fit(X_train_flat, Y_train, batch_size=32, nb_epoch=10, verbose=1, validation_data=(X_val_flat, Y_val))

(26270, 3072) (26270, 43)
(12939, 3072) (12939, 43)
Train on 26270 samples, validate on 12939 samples
Epoch 1/10
26270/26270 [==============================] - 22s - loss: 0.2526 - acc: 0.9242 - val_loss: 0.2819 - val_acc: 0.9240
Epoch 2/10
26270/26270 [==============================] - 13s - loss: 0.2609 - acc: 0.9263 - val_loss: 0.3239 - val_acc: 0.9178
Epoch 3/10
26270/26270 [==============================] - 13s - loss: 0.2207 - acc: 0.9372 - val_loss: 0.3045 - val_acc: 0.9140
Epoch 4/10
26270/26270 [==============================] - 11s - loss: 0.1935 - acc: 0.9424 - val_loss: 0.3022 - val_acc: 0.9158
Epoch 5/10
26270/26270 [==============================] - 10s - loss: 0.2292 - acc: 0.9338 - val_loss: 0.3235 - val_acc: 0.9139
Epoch 6/10
26270/26270 [==============================] - 10s - loss: 0.2034 - acc: 0.9403 - val_loss: 0.3151 - val_acc: 0.9182
Epoch 7/10
26270/26270 [==============================] - 10s - loss: 0.1801 - acc: 0.9474 - val_loss: 0.2637 - val_acc: 0.9304
Ep

In [23]:
print(history.history['acc'])

# accuracy of the last epoch
print(history.history['acc'][-1])

print(history.history['val_acc'])

# accuracy of the last epoch
print(history.history['val_acc'][-1])

[0.92417205939240266, 0.9263037685618275, 0.93722877804662141, 0.94240578604053082, 0.93380281687418132, 0.94031214313812028, 0.94739246288995849, 0.94446136274399484, 0.94084507039077026, 0.9519984773596657]
0.95199847736
[0.92402813202705181, 0.91784527397328974, 0.91398098773460257, 0.91575855941711282, 0.91390370200927606, 0.91823170259992237, 0.93036556150382743, 0.90748898677953438, 0.9089574155607385, 0.91173970170013319]
0.9117397017


In [16]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
assert(history.history['acc'][-1] > 0.92), "The training accuracy was: %.3f" % history.history['acc'][-1]
assert(history.history['val_acc'][-1] > 0.9), "The validation accuracy is: %.3f" % history.history['val_acc'][-1]

**Validation Accuracy**: (0.911)

## Congratulations
You've built a feedforward neural network in Keras!

Don't stop here! Next, you'll add a convolutional layer to drive.py.

## Convolutions
Build a new network, similar to your existing network. Before the hidden layer, add a 3x3 [convolutional layer](https://keras.io/layers/convolutional/#convolution2d) with 32 filters and valid padding.

Then compile and train the network.

Hint 1: The Keras example of a [convolutional neural network](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py) for MNIST would be a good example to review.

Hint 2: Now that the first layer of the network is a convolutional layer, you no longer need to reshape the input images before passing them to the network. You might need to reload your training data to recover the original shape.

Hint 3: Add a [`Flatten()` layer](https://keras.io/layers/core/#flatten) between the convolutional layer and the fully-connected hidden layer.

In [27]:
# TODO: Re-construct the network and add a convolutional layer before the first fully-connected layer.

from keras.layers import Convolution2D, Flatten

model = Sequential()
model.add(Convolution2D(32, 3, 3, activation='relu', border_mode='valid', input_shape=(32,32,3)))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(43, activation='softmax'))
model.summary()

# TODO: Compile and train the model here.
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

history = model.fit(X_train, Y_train, batch_size=128, nb_epoch=20,
          verbose=1, validation_data=(X_val, Y_val))


____________________________________________________________________________________________________
Layer (type)                     Output Shape          Param #     Connected to                     
convolution2d_2 (Convolution2D)  (None, 30, 30, 32)    896         convolution2d_input_2[0][0]      
____________________________________________________________________________________________________
flatten_2 (Flatten)              (None, 28800)         0           convolution2d_2[0][0]            
____________________________________________________________________________________________________
dense_5 (Dense)                  (None, 128)           3686528     flatten_2[0][0]                  
____________________________________________________________________________________________________
dense_6 (Dense)                  (None, 43)            5547        dense_5[0][0]                    
Total params: 3692971
_____________________________________________________________________

In [28]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
assert(history.history['val_acc'][-1] > 0.95), "The validation accuracy is: %.3f" % history.history['val_acc'][-1]

**Validation Accuracy**: (0.9794)

## Pooling
Re-construct your network and add a 2x2 [pooling layer](https://keras.io/layers/pooling/#maxpooling2d) immediately following your convolutional layer.

Then compile and train the network.

In [29]:
# TODO: Re-construct the network and add a pooling layer after the convolutional layer.
from keras.layers import MaxPooling2D

model = Sequential()

model.add(Convolution2D(32, 3, 3, activation='relu', border_mode='valid', input_shape=(32,32,3)))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(43, activation='softmax'))
model.summary()

# TODO: Compile and train the model here.
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

history = model.fit(X_train, Y_train, batch_size=128, nb_epoch=20,
          verbose=1, validation_data=(X_val, Y_val))

____________________________________________________________________________________________________
Layer (type)                     Output Shape          Param #     Connected to                     
convolution2d_3 (Convolution2D)  (None, 30, 30, 32)    896         convolution2d_input_3[0][0]      
____________________________________________________________________________________________________
maxpooling2d_1 (MaxPooling2D)    (None, 15, 15, 32)    0           convolution2d_3[0][0]            
____________________________________________________________________________________________________
flatten_3 (Flatten)              (None, 7200)          0           maxpooling2d_1[0][0]             
____________________________________________________________________________________________________
dense_7 (Dense)                  (None, 128)           921728      flatten_3[0][0]                  
___________________________________________________________________________________________

In [30]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
assert(history.history['val_acc'][-1] > 0.95), "The validation accuracy is: %.3f" % history.history['val_acc'][-1]

**Validation Accuracy**: (0.9774)

## Dropout
Re-construct your network and add [dropout](https://keras.io/layers/core/#dropout) after the pooling layer. Set the dropout rate to 50%.

In [32]:
# TODO: Re-construct the network and add dropout after the pooling layer.

from keras.layers import Dropout
model = Sequential()

model.add(Convolution2D(32, 3, 3, activation='relu', border_mode='valid', input_shape=(32,32,3)))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Dropout(0.5))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(43, activation='softmax'))
model.summary()

# TODO: Compile and train the model here.
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

history = model.fit(X_train, Y_train, batch_size=128, nb_epoch=20,
          verbose=1, validation_data=(X_val, Y_val))


____________________________________________________________________________________________________
Layer (type)                     Output Shape          Param #     Connected to                     
convolution2d_4 (Convolution2D)  (None, 30, 30, 32)    896         convolution2d_input_4[0][0]      
____________________________________________________________________________________________________
maxpooling2d_2 (MaxPooling2D)    (None, 15, 15, 32)    0           convolution2d_4[0][0]            
____________________________________________________________________________________________________
dropout_1 (Dropout)              (None, 15, 15, 32)    0           maxpooling2d_2[0][0]             
____________________________________________________________________________________________________
flatten_4 (Flatten)              (None, 7200)          0           dropout_1[0][0]                  
___________________________________________________________________________________________

In [33]:
# STOP: Do not change the tests below. Your implementation should pass these tests.
assert(history.history['val_acc'][-1] > 0.95), "The validation accuracy is: %.3f" % history.history['val_acc'][-1]

**Validation Accuracy**: (0.9802)

## Optimization
Congratulations! You've built a neural network with convolutions, pooling, dropout, and fully-connected layers, all in just a few lines of code.

Have fun with the model and see how well you can do! Add more layers, or regularization, or different padding, or batches, or more training epochs.

What is the best validation accuracy you can achieve?

**Best Validation Accuracy:** (fill in here)

## Testing
Once you've picked out your best model, it's time to test it.

Load up the test data and use the [`evaluate()` method](https://keras.io/models/model/#evaluate) to see how well it does.

Hint 1: After you load your test data, don't forget to normalize the input and one-hot encode the output, so it matches the training data.

Hint 2: The `evaluate()` method should return an array of numbers. Use the `metrics_names()` method to get the labels.

In [34]:
with open('./test.p', mode='rb') as f:
    test = pickle.load(f)
    
X_test = test['features']
y_test = test['labels']
X_test = X_test.astype('float32')
X_test /= 255
X_test -= 0.5
Y_test = np_utils.to_categorical(y_test, 43)

score = model.evaluate(X_test, Y_test)
print("test score:", score[0])
print("test accuracy:", score[1])

12630/12630 [==============================] - 14s    
test score: 0.473330286238
test accuracy: 0.906809184472


**Test Accuracy:** (0.9068)

## Summary
Keras is a great tool to use if you want to quickly build a neural network and evaluate performance.